docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+    }
+
+    struct FieldLayout
+    {
+        public uint Offset;
+        public FieldType Type;
+    }
+
+    interface IAlgorithmContract
+    {
+        void Init();
+    }
+
+    interface IContract
+    {
+        string Name { get; }
+        uint Version { get; }
+    }
+
+    class Target
+    {
+        sbyte ReadInt8(TargetPointer pointer);
+        byte ReadUInt8(TargetPointer pointer);
+        short ReadInt16(TargetPointer pointer);
+        ushort ReadUInt16(TargetPointer pointer);
+        int ReadInt32(TargetPointer pointer);
+        uint ReadUInt32(TargetPointer pointer);
+        long ReadInt64(TargetPointer pointer);
+        ulong ReadUInt64(TargetPointer pointer);
+        TargetPointer ReadTargetPointer(TargetPointer pointer);
+        byte[] ReadByteArray(TargetPointer pointer, ulong size);
+        void FillByteArray(TargetPointer pointer, byte[] array, ulong size);
+
+        bool TryReadInt8(TargetPointer pointer, out sbyte value);
+        bool TryReadUInt8(TargetPointer pointer, out byte value);
+        bool TryReadInt16(TargetPointer pointer, out short value);
+        bool TryReadUInt16(TargetPointer pointer, out ushort value);
+        bool TryReadInt32(TargetPointer pointer, out int value);
+        bool TryReadUInt32(TargetPointer pointer, out uint value);
+        bool TryReadInt64(TargetPointer pointer, out long value);
+        bool TryReadUInt64(TargetPointer pointer, out ulong value);
+        bool TryReadTargetPointer(TargetPointer pointer, out TargetPointer value);
+        bool TryReadByteArray(TargetPointer pointer, ulong size, out byte[] value);
+        bool TryFillByteArray(TargetPointer pointer, byte[] array, ulong size);
+
+        // If pointer is 0, then the return value will be 0
+        TargetPointer GetTargetPointerForField(TargetPointer pointer, FieldLayout fieldLayout);
+
+        sbyte ReadGlobalInt8(string globalName);
+        byte ReadGlobalUInt8(string globalName);
+        short ReadGlobalInt16(string globalName);
+        ushort ReadGlobalUInt16(string globalName);
+        int ReadGlobalInt32(string globalName);
+        uint ReadGlobalUInt32(string globalName);
+        long ReadGlobalInt64(string globalName);
+        ulong ReadGlobalUInt64(string globalName);
+        TargetPointer ReadGlobalTargetPointer(string globalName);
+
+        bool TryReadGlobalInt8(string globalName, out sbyte value);
+        bool TryReadGlobalUInt8(string globalName, out byte value);
+        bool TryReadGlobalInt16(string globalName, out short value);
+        bool TryReadGlobalUInt16(string globalName, out ushort value);
+        bool TryReadGlobalInt32(string globalName, out int value);
+        bool TryReadGlobalUInt32(string globalName, out uint value);
+        bool TryReadGlobalInt64(string globalName, out long value);
+        bool TryReadGlobalUInt64(string globalName, out ulong value);
+        bool TryReadGlobalTargetPointer(string globalName, out TargetPointer value);
```

Comment:
It might be better to postulate a `GlobalValues,0` contract.

Reply:
I'm trying to keep the contracts such that it's unlikely for work done on one part of the runtime to affect contracts done by somebody else on a different part of the runtime, but I don't have a really strong opinion here. I'd like to hear consensus in the room about this sort of thing.